Mathlib/NumberTheory/FactorisationProperties.lean

```diff
@@ -0,0 +1,320 @@
+    right
+    intro s hs
+    fin_cases hs <;> decide
+
+theorem divisor_subseteq_mul (hn : 0 < n) (hm : 0 < m) :
+    divisors n ∪ divisors m ⊆ divisors (n * m) := by
+  intro a ha
+  simp only [mem_union, mem_divisors] at ha ⊢
+  rcases ha with hn1 | hm1
+  · exact ⟨dvd_trans hn1.1 ⟨m, rfl⟩, by positivity⟩
+  · exact ⟨dvd_trans hm1.1 ⟨n, mul_comm ..⟩, by positivity⟩
+
+theorem prop_divisors_subseteq_mul (hn : 0 < n) (hm : 0 < m) :
+    properDivisors n ∪ properDivisors m ⊆ properDivisors (n*m) := by
+      refine subset_iff.mpr ?_
+      rintro a ha
+      have ha' : a ∈ properDivisors n ∨ a ∈ properDivisors m := by
+        simp_all only [gt_iff_lt, le_mul_iff_one_le_right,
+          le_mul_iff_one_le_left, mem_union, mem_properDivisors]
+      rcases ha' with hn1 | hm1
+      · have han : a ∣ n := by
+          simp_all only [mem_properDivisors, and_self]
+        refine mem_properDivisors.mpr ?_
+        constructor
+        exact Dvd.dvd.mul_right han m
+        refine lt_mul_of_lt_of_one_le ?_ hm
+        simp_all only [gt_iff_lt, le_mul_iff_one_le_right, mem_properDivisors, true_and, ne_eq]
+      · have ham : a ∣ m := by
+          simp_all only [gt_iff_lt, le_mul_iff_one_le_right, mem_properDivisors]
+        refine mem_properDivisors.mpr ?_
+        constructor
+        exact Dvd.dvd.mul_left ham n
+        refine lt_mul_of_one_le_of_lt hn ?_
+        simp_all only [gt_iff_lt, le_mul_iff_one_le_right, le_mul_iff_one_le_left,
+          mem_properDivisors, true_and, ne_eq]
+
+lemma divisors_eq_proper_union_self (hn : 0 < n) :
+    divisors n = properDivisors n ∪ {n} := by
+      dsimp [divisors, properDivisors]
+      ext a
+      simp [*]
+      constructor
+      <;> rintro ⟨⟨h1, h2⟩, h3⟩
+      · omega
+      · omega
+      · simp only [*, lt_add_iff_pos_right, zero_lt_one, and_true, dvd_refl]
+        omega
+
+lemma deficient_not_abundant_or_perfect (hn : 0 < n) :
+    Deficient n ↔ ¬ Abundant n ∧ ¬ Perfect n := by
+      dsimp [Perfect, Abundant, Deficient]
+      omega
+
+lemma perfect_not_abundant_or_deficient (hn : 0 < n) :
+    Perfect n ↔ ¬ Abundant n ∧ ¬ Deficient n := by
+      dsimp [Perfect, Abundant, Deficient]
+      omega
+
+lemma abundant_not_perfect_or_deficient (hn : 0 < n) :
+    Abundant n ↔ ¬ Perfect n ∧ ¬ Deficient n := by
+      dsimp [Perfect, Abundant, Deficient]
+      omega
+
+/- All numbers are either deficient, perfect, or abundant -/
+theorem deficient_or_perfect_or_abundant (hn : 0 < n) :
+    Deficient n ∨ Abundant n ∨ Perfect n := by
+      dsimp [Perfect, Abundant, Deficient]
+      omega
+
+theorem perfect_pseudoperfect (h : Perfect n) : Pseudoperfect n := by
+  rcases h with ⟨h1, h2⟩
+  constructor
+  · exact h2
+  · use properDivisors n
+    constructor
+    · exact mem_powerset_self (properDivisors n)
+    · exact h1
+
+theorem prime_not_abundant (h : Prime n) : ¬ Abundant n := by
+  intro h1
+  have h2 : ∑ i in properDivisors n, i = 1 := by exact sum_properDivisors_eq_one_iff_prime.mpr h
```

Comment:
Can you avoid the `by exact` here?
Does `omega` require all these hypotheses?

Reply:
I condensed the three hypotheses into two. The omega does require the first and the last ones to complete the goal.